test/server/config_validation/xds_fuzz.cc

```diff
@@ -184,77 +186,244 @@ void XdsFuzzTest::replay() {
       }
 
-      if (sent_listener) {
-        EXPECT_TRUE(
-            waitForAck(Config::TypeUrl::get().RouteConfiguration, std::to_string(version_)));
-      }
+      EXPECT_TRUE(waitForAck(Config::TypeUrl::get().RouteConfiguration, std::to_string(version_)));
+
+      ENVOY_LOG_MISC(info, "added: {}, modified {}, removed {}, warming {}, active {}",
+                     test_server_->counter("listener_manager.listener_added")->value(),
+                     test_server_->counter("listener_manager.listener_modified")->value(),
+                     test_server_->counter("listener_manager.listener_removed")->value(),
+                     test_server_->gauge("listener_manager.total_listeners_warming")->value(),
+                     test_server_->gauge("listener_manager.total_listeners_active")->value());
+
       break;
     }
     case test::server::config_validation::Action::kRemoveRoute: {
+      ENVOY_LOG_MISC(info, "Ignoring request to remove route_{}",
+                     action.remove_route().route_num());
+      // it seems like routes cannot be removed
+      // leaving a route out of an SOTW request does not remove it and sending a
+      // remove message in a delta request is ignored
+      break;
+
       if (sotw_or_delta_ == Grpc::SotwOrDelta::Sotw) {
         // routes cannot be removed in SOTW updates
+        ENVOY_LOG_MISC(info, "Ignoring request to remove route_{}",
+                       action.remove_route().route_num());
         break;
       }
+      ENVOY_LOG_MISC(info, "Removing route_{}", action.remove_route().route_num());
 
       auto removed = removeRoute(action.remove_route().route_num());
+
       if (removed) {
         updateRoute(routes_, {}, {*removed});
-        EXPECT_TRUE(
-            waitForAck(Config::TypeUrl::get().RouteConfiguration, std::to_string(version_)));
+        verifier_.routeRemoved(*removed);
+      } else {
+        updateRoute(routes_, {}, {});
       }
       break;
     }
     default:
       break;
     }
+
+    // TODO(samflattery): makeSingleRequest here?
+    version_++;
   }
 
+  verifyState();
   close();
 }
 
+void XdsFuzzTest::drainListener(const std::string& name) {
+  // ensure that the listener drains correctly by checking that it is still draining halfway through
+  ENVOY_LOG_MISC(info, "Draining {}", name);
+  /* EXPECT_EQ(test_server_->gauge("listener_manager.total_listeners_draining")->value(), */
+  /*           verifier_.numDraining()); */
+  /* timeSystem().advanceTimeWait(std::chrono::milliseconds(300000)); */
+  /* EXPECT_EQ(test_server_->gauge("listener_manager.total_listeners_draining")->value(), */
+  /*           verifier_.numDraining()); */
+  /* timeSystem().advanceTimeWait(std::chrono::milliseconds(300001)); */
+  /* EXPECT_EQ(test_server_->gauge("listener_manager.total_listeners_draining")->value(), */
+  /*           verifier_.numDraining() - 1); */
+  verifier_.drainedListener(name);
+}
+
+void XdsFuzzTest::verifyListeners() {
```

Comment:
what do you think about factoring a lot of this logic into the verifier? 

as in, we pass in the listener rep's name and state to the verifier, and the verifier holds the logic in here that compares?

Reply:
I was thinking of doing this, but then I would need to pass `test_server` in to the verifier class as well to be able to check stats, etc, and `test_server_` is a unique_ptr AFAIK so I'm not sure if I'm able to do this?